### PR TITLE
[Bugfix][CPU][Spec Decode] Write expand_kernel output at caller dtype in CPU shim

### LINKED ISSUE #55005: [Bug][CPU][Spec Decode]: CPU expand_kernel shim discards its output — non-greedy spec decode consumes uninitialized memory as temperature/top_k/top_p
### Summary

The CPU monkey-patch for `expand_kernel` (`vllm/utils/cpu_triton_utils.py:408-421`) passes `_ensure_int64(output)` to the int64-only C++ kernel (`csrc/cpu/spec_decode_utils.cpp:405-428`). When the caller's buffer is not int64, `_ensure_int64` returns a **new tensor** (`cpu_triton_utils.py:55-56`); the kernel fills that temporary and the caller's buffer is never written. Every sibling wrapper in the file tracks `orig_dtype` and copies back — this one is the only one that doesn't.

Every production call hits it: `expand_batch_to_tokens` allocates `x.new_empty(num_tokens)` in x's dtype (`vllm/v1/sample/rejection_sampler.py:596`) and the three production inputs are float32 temperature, int32 top_k, float32 top_p (`vllm/v1/worker/gpu_input_batch.py:199-216`, call sites `rejection_sampler.py:537-560`). So on any CPU spec-decode run with a non-greedy request, `logits.div_(temperature)` (`:545`) and `apply_top_k_top_p` (`:565`) consume **uninitialized `new_empty` memory** — fresh zero pages give `div_(0)` → inf logits; arbitrary garbage gives the out-of-bounds crash previously observed downstream. A second layer: even with copy-back, `_ensure_int64(input_val)` truncates floats — temperature 0.7 → 0 → rewritten to 1 by the `replace_from=0` logic.

This path is live whenever Triton is absent (`cpu_model_runner.py:65-73` skips the patches only when `HAS_TRITON`; :115 installs this one) — i.e., pip installs without triton-cpu, macOS/aarch64.

### This was reported before, and the fix everyone believed in is the buggy code

PR #42638 reported exactly this corruption (uninitialized `expand_batch_to_tokens` output; third-party-confirmed on Xeon — OOB crash on Qwen's 151k vocab, **silently in-bounds on Llama's 128k**) via a different mechanism: the raw Triton kernel no-oping with no driver. A maintainer redirected the fix to the CPU `expand_kernel_impl` monkey-patch, and the PR was closed after #44991 (merged) fixed the Triton-CPU path, on the stated premise that the "existing C++ monkey-patch fallbacks" cover non-Triton environments. That premise was false: the expand fallback has never produced correct output for any of the three production dtypes.

### Reproduction (seconds; no model, no server, no GPU)

The script below, on a from-source CPU build at `605c3ddc` (`vllm 0.28.1rc1.dev315+g605c3ddcb.cpu`, torch 2.13.0, macOS arm64):

```
(a) int64 control      -> [10, 10, 20, 20, 20, 30]  expected [10, 10, 20, 20, 20, 30]  OK
(b) float32 out buffer -> [-777.0, -777.0, -777.0, -777.0, -777.0, -777.0]  UNTOUCHED (output discarded)
(c) fp32 temperature   -> [0.0, 0.0, 0.0, 0.0, 0.0, 0.0]
    expected              [0.699999988079071, 0.699999988079071, 0.5, 0.5, 0.5, 0.8999999761581421]  DIVERGES (uninitialized new_empty)
(d) fp32 in, i64 out   -> [1, 1, 1, 1, 1, 1]  (0.7/0.5/0.9 all truncated to 0, then 'replaced' with 1)
(e) int32 top_k        -> [0, 0, 0, 0, 0, 0]
    expected              [5, 5, 40, 40, 40, 1]  DIVERGES
```

Arm (b) is deterministic proof: a sentinel-filled output buffer comes back untouched. Arm (d) isolates the truncation layer.

<details>
<summary>repro_expand_kernel_cpu.py</summary>

```python
# Repro: the CPU monkey-patch for `expand_kernel` discards its output and
# truncates float inputs — CPU spec decode with any non-greedy request
# consumes uninitialized memory as temperature / top_k / top_p.
#
# `vllm/utils/cpu_triton_utils.py::_expand_kernel_impl` passes
# `_ensure_int64(output)` to the C++ kernel. When the caller's buffer is not
# int64 (production passes float32 temperature/top_p and int32 top_k),
# `_ensure_int64` allocates a NEW tensor; the C++ kernel fills that temporary
# and the caller's `new_empty` buffer is returned uninitialized. Every sibling
# wrapper in the file copies back; this one doesn't. A second layer: float
# inputs are truncated to int64 (temperature 0.7 -> 0 -> replaced with 1).
#
# No model, no server, no GPU; seconds. Run on a CPU build of vLLM
# (path is live when Triton is absent, i.e. the C++ monkey-patch path).

import torch

import vllm  # noqa: F401
import vllm._C  # noqa: F401  (registers the torch.ops._C CPU kernels)
import vllm.utils.cpu_triton_utils as cpu_tl
import vllm.v1.sample.rejection_sampler as rs

# Monkey-patch exactly as vllm/v1/worker/cpu_model_runner.py:115 does when
# Triton is unavailable.
rs.expand_kernel = cpu_tl.expand_kernel

print(f"torch {torch.__version__}")
from importlib.metadata import version  # noqa: E402

print(f"vllm  {version('vllm')}")
print()

cu = torch.tensor([2, 5, 6], dtype=torch.int64)  # 3 requests -> 6 tokens

# --- Arm (a): int64 end-to-end (control: the C++ kernel itself is fine) ---
x = torch.tensor([10, 20, 30], dtype=torch.int64)
out = rs.expand_batch_to_tokens(x, cu, num_tokens=6)
print(f"(a) int64 control      -> {out.tolist()}  expected [10, 10, 20, 20, 20, 30]"
      f"  {'OK' if out.tolist() == [10, 10, 20, 20, 20, 30] else 'FAIL'}")

# --- Arm (b): output buffer is provably never written (deterministic) ---
# Call the shim directly with a float32 output buffer pre-filled with a
# sentinel. If the kernel wrote through, the sentinel would be gone.
out_f32 = torch.full((6,), -777.0, dtype=torch.float32)
cpu_tl._expand_kernel_impl(out_f32, torch.tensor([0.7, 0.5, 0.9]), cu, 0, 1)
untouched = bool((out_f32 == -777.0).all())
print(f"(b) float32 out buffer -> {out_f32.tolist()}  "
      f"{'UNTOUCHED (output discarded)' if untouched else 'written'}")

# --- Arm (c): the production temperature path (float32, new_empty) ---
# expand_batch_to_tokens allocates `x.new_empty(num_tokens)` in x's dtype;
# with the output discarded, callers receive uninitialized memory. This is
# exactly rejection_sampler.py:537-545: the result divides the logits.
temp = torch.tensor([0.7, 0.5, 0.9], dtype=torch.float32)
got = rs.expand_batch_to_tokens(temp, cu, num_tokens=6, replace_from=0, replace_to=1)
expected = torch.repeat_interleave(temp, torch.tensor([2, 3, 1]))
print(f"(c) fp32 temperature   -> {got.tolist()}")
print(f"    expected              {expected.tolist()}  "
      f"{'DIVERGES (uninitialized new_empty)' if not torch.equal(got, expected) else 'ok'}")

# --- Arm (d): float truncation, isolated from the discard bug ---
# With an int64 output buffer `_ensure_int64(output)` is the identity, so the
# kernel's writes ARE visible — and show the input-side truncation:
# temperature 0.7 -> int64 0 == replace_from -> replaced with 1.
out_i64 = torch.full((6,), -777, dtype=torch.int64)
cpu_tl._expand_kernel_impl(out_i64, temp, cu, 0, 1)
print(f"(d) fp32 in, i64 out   -> {out_i64.tolist()}  "
      f"(0.7/0.5/0.9 all truncated to 0, then 'replaced' with 1)")

# --- Arm (e): int32 top_k path (production dtype, gpu_input_batch.py:216) ---
topk = torch.tensor([5, 40, 1], dtype=torch.int32)
got_k = rs.expand_batch_to_tokens(topk, cu, num_tokens=6)
exp_k = torch.repeat_interleave(topk, torch.tensor([2, 3, 1]))
print(f"(e) int32 top_k        -> {got_k.tolist()}")
print(f"    expected              {exp_k.tolist()}  "
      f"{'DIVERGES' if not torch.equal(got_k, exp_k) else 'ok'}")
```

</details>

### What I cannot determine

- When the shim was introduced: `git log -S '_expand_kernel_impl'` bottoms out at my clone's graft boundary (depth 3,263), so introduction predates visible history.
- Whether the int64-only C++ kernel was a deliberate scope decision — the wrapper's discard + truncation is a bug either way.

### Fix

PR attached (posted alongside this issue): expand natively at the caller's dtype in the shim — the same `torch.repeat_interleave` semantics #42638 proposed, relocated to the monkey-patch layer where the maintainer asked for it. Regression test fails 3/4 arms before the fix, passes 4/4 after.

### Disclosure

This report was prepared with AI assistance (Claude Code); the analysis, measurements, and fix were reviewed and run by the submitter.

### PR DESCRIPTION
**Pull request by mosafariuk**
## Purpose

Fixes #55005.

The CPU monkey-patch for `expand_kernel` (`vllm/utils/cpu_triton_utils.py`) routed all tensors through `_ensure_int64`, which returns a *new* tensor for non-int64 inputs. The C++ kernel filled that temporary; the caller's buffer — allocated by `expand_batch_to_tokens` as `x.new_empty(num_tokens)` in x's dtype — was never written. Production passes float32 temperature/top_p and int32 top_k, so every CPU spec-decode step with a non-greedy request divided logits by uninitialized memory (zero pages → inf logits; arbitrary garbage → the OOB crash reported in #42638). Float inputs were additionally truncated (temperature 0.7 → 0 → rewritten to 1 by the `replace_from=0` logic).

## Fix

Expand natively at the caller's dtype in the shim: `torch.repeat_interleave` over `diff(cu_num_tokens)` with the `replace_from → replace_to` rewrite applied at input dtype before the store — matching the Triton kernel's compare-before-store-cast ordering. This is the semantics #42638 proposed in `expand_batch_to_tokens`, relocated to the CPU shim layer per the maintainer guidance on that PR (keep hardware-specific handling out of the common code path).

**Not a duplicate:** #42638 (closed unmerged) targeted the no-Triton-driver kernel no-op inside `expand_batch_to_tokens` and was closed after #44991 covered the Triton-CPU path, on the premise that the C++ monkey-patch fallbacks handle non-Triton environments. This PR fixes the fallback itself, which never produced correct output for the production dtypes. No open PR touches `cpu_triton_utils.py` or this path (searched `expand_kernel`, `expand_batch_to_tokens`, `cpu_triton_utils` in open PRs, 2026-09-02).

## Test Plan

New regression test `test_expand_kernel_cpu_shim_writes_output_at_caller_dtype` in `tests/v1/sample/test_rejection_sampler.py`: sentinel-filled output buffers across the three production dtypes plus the int64 control, asserting in-place writes with correct replace semantics and no truncation. Pure-torch shim, so the test runs on any platform.

```
# Before fix (main @ 605c3ddc):
3 failed, 1 passed   # int64 control passes via the dtype-identity path — the exact bug shape
# After fix:
4 passed
```

Baseline attribution: the rest of `tests/v1/sample/test_rejection_sampler.py` fails identically (75) on unmodified main in this no-GPU environment (the file targets the Triton path and is not in CPU CI); this PR changes none of those results.

Direct-shim repro (also in the linked issue) measured before/after on a from-source CPU build at `605c3ddc`: sentinel buffer untouched before, correct expansion after; fp32 temperature `[0,0,0,0,0,0]` → `[0.7, 0.7, 0.5, 0.5, 0.5, 0.9]`; int32 top_k `[0,...]` → `[5, 5, 40, 40, 40, 1]`.

Lint: `pre-commit run --files vllm/utils/cpu_triton_utils.py tests/v1/sample/test_rejection_sampler.py` — all hooks pass (ruff check/format, mypy-3.10, SPDX).

Model evals: not run — no GPU available to this contributor; the change affects only the CPU fallback path and restores the documented expand semantics verified by the unit test above.

## Disclosure

AI assistance (Claude Code) was used for the audit, repro, and patch; every changed line was reviewed and all commands above were run by the submitter. Commit carries a `Co-authored-by` trailer per AGENTS.md.

## Summary by CodeRabbit

* **Bug Fixes**
  * Fixed CPU sampling expansion to preserve the caller’s original data types.
  * Prevented truncation of floating-point values during replacement and expansion operations.
  * Ensured results are written directly to the provided output, improving correctness for CPU execution.

* **Tests**
  * Added coverage for floating-point and integer sampling inputs, including in-place output behavior.

### CODE REVIEW

**Review by Manny7717 — APPROVED** [review]
Fix verified — regression-proven and semantically sound.

**Bug confirmed on base:** the CPU shim routed every tensor through `_ensure_int64`, so a non-int64 `output` was an int64 *copy* the kernel wrote into and never copied back — the caller's buffer (allocated at caller dtype in `expand_batch_to_tokens`, `x.new_empty(num_tokens)`) stayed uninitialized, and float inputs (temperature 0.7 etc.) were truncated to int before the replace comparison. Executed: the 4 new parametrized cases FAIL on base code (here the compiled `torch.ops._C.expand_kernel_impl` doesn't even exist in this CPU build, so the shim raises AttributeError — a hard crash of the same path) and PASS on head.

**Fix correct:** the torch-native implementation is exactly the documented contract — `expand_batch_to_tokens`'s own docstring example (`x=[a,b,c]`, `cu=[2,5,6]` → `[a,a,b,b,b,c]`) is literally `torch.diff(cu, prepend=zeros(1))` counts + `repeat_interleave`, with the `replace_from→replace_to` rewrite applied per token value before expansion (equivalent to the C++ kernel's per-segment copy since all copies of a token share the value). dtype preserved end-to-end via `torch.where`/`full_like`/`repeat_interleave` at the input dtype; `output[:n] = expanded` writes through to the caller's buffer in place. GPU/triton path untouched (separate real-kernel implementation); `MAX_NUM_TOKENS` was already unused in the impl body. Production reachability confirmed: `apply_sampling_constraints` (rejection_sampler.py:537-557) calls `expand_batch_to_tokens` with per-request float32 temperature/top_p and int32 top_k on the CPU path — exactly the failing shapes.

Lint: ruff check + format clean on both files. No competing PRs (checked `is:pr` for expand_kernel/CPU spec decode).

**Comment by mosafariuk** [comment]
Thanks for the thorough verification — running both directions independently is exactly the check this needed. The AttributeError data point is a useful addition: on builds where the compiled kernel op is absent entirely, the old path was a hard crash rather than silent corruption, which makes the blast radius wider than the issue described. Good to have that on the record.
